mithril-client/src/cardano_database_client/download_unpack.rs

```diff
@@ -200,229 +216,189 @@ impl InternalArtifactDownloader {
-                    );
-                }
-            }
-        }
-
-        Ok(immutable_file_numbers_downloaded)
-    }
-
-    fn spawn_immutable_download_future(
-        &self,
-        args: DownloadImmutableFutureBuilderArgs,
-    ) -> MithrilResult<Pin<Box<DownloadImmutableFuture>>> {
-        let logger_clone = self.logger.clone();
-        let download_id_clone = args.download_id.to_string();
-        let file_downloader = args.file_downloader;
-        let file_downloader_uri = args.file_downloader_uri;
-        let compression_algorithm = args.compression_algorithm;
-        let immutable_files_target_dir = args.immutable_files_target_dir;
-        let immutable_file_number = args.immutable_file_number;
-        let file_size = args.file_size;
-        let download_future = async move {
-            let downloaded = file_downloader
-                .download_unpack(
-                    &file_downloader_uri,
-                    file_size,
-                    &immutable_files_target_dir,
+    ) -> MithrilResult<DownloadTask> {
+        let mut locations_to_try = vec![];
+        let mut locations_sorted = locations.to_owned();
+        locations_sorted.sort();
+        for location in locations_sorted {
+            let location_to_try = match location {
+                ImmutablesLocation::CloudStorage {
+                    uri,
                     compression_algorithm,
-                    DownloadEvent::Immutable {
-                        immutable_file_number,
-                        download_id: download_id_clone.clone(),
-                    },
-                )
-                .await;
-            match downloaded {
-                Ok(_) => Ok(immutable_file_number),
-                Err(e) => {
-                    slog::error!(
-                        logger_clone,
-                        "Failed downloading and unpacking immutable file {immutable_file_number} for location {file_downloader_uri:?}"; "error" => ?e
+                } => {
+                    let file_downloader = self.http_file_downloader.clone();
+                    let file_downloader_uri = FileDownloaderUri::from(
+                        uri.expand_for_immutable_file_number(immutable_file_number),
                     );
-                    Err(e.context(format!("Failed downloading and unpacking immutable file {immutable_file_number} for location {file_downloader_uri:?}")))
-                }
-            }
-        };
 
-        Ok(Box::pin(download_future))
-    }
+                    LocationToDownload {
+                        file_downloader,
+                        file_downloader_uri,
+                        compression_algorithm: compression_algorithm.to_owned(),
+                    }
+                }
+                ImmutablesLocation::Unknown => {
+                    return Err(anyhow!("Unknown location type to download immutable"));
```

Comment:
If `Unknown` is met, we probably want to skip to next location and return `None` (while returning `Some` in other cases). This will allow to push to the `location_to_try` only if the value is something, and return an error only if the `location_to_try` list is empty.

This also applies to the ancillary task creation.

Reply:
The PR have been modified to add a 'sanitize' step before using the locations.
If all locations are unknown an error is raised, else unknown locations are removed from the list, before even reaching those `match` (hence they are now marked as `unreachable!()`).